Addition of test for finaliser callback with major cycle

ctk21 opened the pull request:
This is a follow up to #530 which adds a test where a finaliser is run with a root in a register; which was the source of the bug in #517. 

This is the same bug, but is slightly different in how it manifests because the register root is missed by the major collector when the finaliser forces a complete cycling of the heap.

ctk21 committed:
Addition of test for finaliser callbacks while block is held in a register

sadiqj commented on testsuite/tests/callback/test_finaliser_gc.ml:
```diff
@@ -0,0 +1,31 @@
+(* TEST
+*)
+
+let rec tak (x, y, z as _tuple) =
```
This function is redundant.

sadiqj commented on testsuite/tests/callback/test_finaliser_gc.ml:
```diff
@@ -0,0 +1,31 @@
+  (* Spin on the minor heap allocating but keep [s] in a
+    register and force a major cycle such that the
+    finaliser runs *)
+  for x = 0 to 10_000_000 do
```
Presumably this is dependent on the size of the minor heap. If I set a minor heap a bit larger than 640mb I think this will fail as well?

Maybe we should use the current Gc minor heap size to set the iteration count.

ctk21 replied:
I've made this less dependent on the minor heap size and also it will complete a bit quicker when things work as expected. I flag when the finaliser has run to break the loop. The upper bound is there to prevent the test hanging in the (hopefully unlikely) case that the finaliser code gets broken. 

The upper bound will prevent you having arbitrarily large minor heaps; we are allocating of order 8Bn words if this doesn't force the finaliser to run then the test will not work as expected. 

I've verified that the changes keep things in the register and correctly fail on the old buggy code before #530.

ctk21 committed:
Remove unused tak.
Improve termination condition of test_finaliser_gc to handle larger minor heaps and also exit quicker when success
